### PR TITLE
Feat/infinite scroll

### PR DESCRIPTION
**Pull request by SergeiPleshkov**
Implements #60 

## Description
Implemented infinite scroll component, added docs
## Markup:

<details>

```vue
  <div
    class="va-infinite-scroll"
    :class="{reverse: reverse}"
  >
    <slot
      name="default"
      ref="content"
    />
    <slot
      name="loading"
      v-if="fetching"
    >
      <div>Loading...</div>
    </slot>
  </div>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

### CODE REVIEW

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+</script>
+
+<style lang="scss">
+@import "../../vuestic-sass/global/utility-global";
+
+.va-infinite-scroll {
+  display: flex;
+  flex-direction: column;
+
+  &.reverse {
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: bem

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.demo.vue [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      setTimeout(() => {
+        this.reverseRecords.unshift(...this.getNewRecords())
+        done()
+      }, 2000)
+    },
+  },
+}
+</script>
+<style lang="scss">
+.scroll__container {
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.demo.vue]: Better use scoped or inline styles.

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    debounce: {
+      type: Number,
+      required: false,
+      default: 100,
+    },
+  },
+  data () {
+    return {
+      index: 0,
+      fetching: false,
```
[the remark this answers is not in the document]
Comment (SergeiPleshkov) [reply src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: fetching - state between requesting and receiving data
working - state with active event listeners, no matter if it is fetching now or not (see watcher for disabled)

[the remark this answers is not in the document]
Comment (SergeiPleshkov) [reply src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: Seems like not necessary, changed checks to disabled.

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.demo.vue [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                :size="48"
+                :color="$themes.primary"
+              />
+            </template>
+          </va-infinite-scroll>
+        </div>
+      </VbCard>
+
+      <VbCard>
+        <div class="scroll__container query">
```
[the remark this answers is not in the document]
Comment (SergeiPleshkov) [reply src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.demo.vue]: It is used as a scroll-target query selector.

**Comment by asvae** [comment]
We want to use `:load` instead `load`. And use promise based handling (please also test case where request errors out).

<VbCard
        title="promise based api"
      >
        <va-infinite-scroll
          disabled
          :load="appendRecordsAsync"
        >
          <ul>
            <li
              v-for="(record, index) in disabledRecords"
              :key="record.id"
            >
              {{ record.text }} #{{ index }}
            </li>
          </ul>
        </va-infinite-scroll>
      </VbCard>
(edited)

6:57
 async appendRecordsAsync () {
      await sleep(20000)
      this.reverseRecords.unshift(...this.getNewRecords())
    },

**Comment by asvae** [comment]
It appear loader is shown only after debounce interval passes. Probably should be shown right after users scrolls down.

**Comment by asvae** [comment]
.spinner class - use bem instead.

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+import * as _ from 'lodash'
+import VaProgressCircle from '../va-progress-bar/progress-types/VaProgressCircle'
+import { makeContextablePropsMixin } from '../../context-test/context-provide/ContextPlugin'
+import { sleep } from '../../../services/utils'
+
+export default {
+  name: 'VaInfiniteScroll',
+  components: { VaProgressCircle },
+  mixins: [
+    makeContextablePropsMixin({
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: Would be great to test context (see `testIsContextableComponent`).
+ add snapshot for default case (that way you can get breaking test on `this.$theme.danger` for instance). See button (it has too many snapshots though, we only want one case).

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+      offset: {
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: Any reason to have default offset at 1? Does it matter if we use 0? Also I see quasar is using 500 as default. Why won't we?

Might also make sense to double-check with refs if loader size is included in offset or not and act accordingly.

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+      if (this.reverse) {
+        const heightDifference =
+        this.$el.offsetHeight - this.initialHeight
+        this.scrollTargetElement.scrollTop = heightDifference
+      }
+    },
+    resume () {
+      if (!this.disabled) {
+
+          this.scrollTargetElement.addEventListener(
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: run lint + double check why it doesn't work for you. You might want to merge from dev for updated deps.

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+  data () {
+    return {
+      index: 0,
+      fetching: false,
+      error: false,
+      initialHeight: null,
+    }
+  },
+  watch: {
+    disabled (value) {
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: That seems to be broken. 
image

File: docs/components/VaInfiniteScroll.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+- [ ] `default` - slot for scrolled content
+- [ ] `loading` - expected: `spinner` - slot displaying something while loading content
+
+# Example of loading method
+
+
+```typescript
+    appendRecordsAsync () {
+      return sleep(2000).then(() => {
```
Comment (asvae) [inline docs/components/VaInfiniteScroll.md]: async/await?
Something like this seems much more readable to me.
```js
await sleep(2000)
this.records.push(...this.getNewRecords())
```

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.demo.vue [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    VaProgressCircle,
+  },
+  data () {
+    return {
+      records: this.getInitialRecords(),
+      disabledRecords: this.getInitialRecords(),
+    }
+  },
+  methods: {
+    getInitialRecords () {
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.demo.vue]: You can declare these methods outside of component, given these are just factories and do not rely on `this`.

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.demo.vue [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+      </VbCard>
+
+      <VbCard title="load failed">
+        <div
+          class="scroll__container"
+        >
+          <va-infinite-scroll
+            :load="failedLoadAsync"
+          >
+            <li
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.demo.vue]: `li` without `ul` is not a valid markup.

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+      ref="content"
+    />
+    <slot
+      name="loading"
+      v-if="!disabled"
+    >
+      <div ref="defaultSpinner"
+           class="va-infinite-scroll__spinner"
+           :class="{'va-infinite-scroll__spinner--invisible': !fetching}">
+        <va-progress-circle
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: Progress circle still changes its size while spinning. You might want to fix this (sounds like 1-2 lines of code fix to me).

File: docs/components/VaInfiniteScroll.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        :color="$themes.primary"
+      />
+    </template>
+  </va-infinite-scroll>
+</div>
+```
+
+# Props
+
+- [ ] `offset: Number` - offset (pixels) to bottom of Infinite Scroll container from which the component should start loading more content in advance
```
Comment (asvae) [inline docs/components/VaInfiniteScroll.md]: Why checkboxes? :)

File: src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+      this.load()
+        .then(this.finishLoading)
+        .catch(this.onError)
+    },
+    onError (e) {
+      this.stop()
+      this.error = true
+      this.fetching = true
+      sleep(2000).then(this.stopErrorDisplay).then(this.resume)
```
Comment (asvae) [inline src/components/vuestic-components/va-infinite-scroll/VaInfiniteScroll.vue]: You may want to discuss this behavior with @xx13. 
Seems too slow to me. I would've expected 0.2-0.3ms blink. Splitting into separate issue could be an option.

**Pull request merged**